libs/elasticsearch-nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java

```diff
@@ -108,14 +126,82 @@ public boolean connect() throws IOException {
         return isConnected;
     }
 
-    public abstract int read() throws IOException;
+    public void sendMessage(Object message, BiConsumer<Void, Throwable> listener) {
```

Comment:
this one could accept WriteOperation and then we could just add the context and the listener to it via setters?